### PR TITLE
Add installation testing option

### PR DESCRIPTION
**Pull request by jjwilke**
Allows easier testing Jenkins of Kokkos installations. 
* Sets all tests to link to alias `Kokkos::kokkos`
* Calls `find_package(Kokkos)` instead of building Kokkos when option is set

### CODE REVIEW

File: .jenkins [hunk 1]
```diff
@@ -122,19 +122,32 @@ pipeline {
...
+                              .. && \
+                              make -j8 install && \
+                              cd .. && \
+                              rm -rf build-tests && mkdir -p build-tests && cd build-tests && \
+                              export CMAKE_PREFIX_PATH=${PWD}/../install && \
+                              cmake \
+                                -DCMAKE_BUILD_TYPE=Release \
+                                -DCMAKE_CXX_COMPILER_LAUNCHER=ccache \
+                                -DCMAKE_CXX_COMPILER=$WORKSPACE/bin/nvcc_wrapper \
+                                -DCMAKE_CXX_FLAGS=-Werror \
```
Comment (jjwilke) [inline .jenkins:148]: I guess this is okay (it works). We could have also done `-DKokkos_ROOT=...`

File: .jenkins [hunk 2]
```diff
@@ -122,19 +122,32 @@ pipeline {
...
-                        sh '''rm -rf build && mkdir -p build && cd build && \
+                        sh '''rm -rf install && mkdir -p install && \
+                              rm -rf build && mkdir -p build && cd build && \
                               cmake \
                                 -DCMAKE_BUILD_TYPE=Release \
                                 -DCMAKE_CXX_COMPILER_LAUNCHER=ccache \
                                 -DCMAKE_CXX_COMPILER=$WORKSPACE/bin/nvcc_wrapper \
                                 -DCMAKE_CXX_FLAGS=-Werror \
                                 -DKokkos_ENABLE_COMPILER_WARNINGS=ON \
-                                -DKokkos_ENABLE_TESTS=ON \
```
Comment (dalg24) [inline .jenkins:132]: Why would we not also run the regular test suite on that build?

File: .jenkins (hunk 1 above)
Comment (masterleinad) [reply .jenkins:148]: The enabled tests depend on the backends enabled here, don't they? Hence, these should match.

File: CMakeLists.txt [hunk 3]
```diff
@@ -147,6 +147,10 @@ INCLUDE(${KOKKOS_SRC_PATH}/cmake/kokkos_tribits.cmake)
 # Check the environment and set certain variables
 # to allow platform-specific checks
 INCLUDE(${KOKKOS_SRC_PATH}/cmake/kokkos_check_env.cmake)
+
+#
+KOKKOS_OPTION(INSTALL_TESTING OFF BOOL "Whether to build tests and examples against installation")
```
Comment (masterleinad) [inline CMakeLists.txt:152]: What this is doing effectively is to disable compiling the library and just run the tests.
Hence, I would rather name it `DISABLE_COMPILING` or similar and make sure `Kokkos_ENABLE_TESTS=ON`.

**Comment by crtrott** [comment]
Ok it doesn't look like this fully works. The config is doing something in between where it doesn't get the enabled spaces from the installed version. It would work if I would just give matching options for the second configure but do we want that?

**Comment by jjwilke** [comment]
@masterleinad Yeah, we should probaly just make it an internal variable.
@crtrott We would need to make the configurations match. We could make this more sophisticated (and also test `kokkos_check`), but that's probably overkill for now.

**Comment by crtrott** [comment]
So I tried something where I stuck the check just before KOKKOS_SETUP_BUILD_ENVIRONMENT in the main CMakeLists:
```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 4ec1bac..24fde2a 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -151,6 +151,17 @@ INCLUDE(${KOKKOS_SRC_PATH}/cmake/kokkos_check_env.cmake)
 #
 KOKKOS_OPTION(INSTALL_TESTING OFF BOOL "Whether to build tests and examples against installation")

+# Short circuit all the rest of the processing if we just build tests against installed Kokkos
+IF(KOKKOS_INSTALL_TESTING)
+  SET(KOKKOS_ENABLE_TESTS ON)
+  KOKKOS_PACKAGE_DECL()
+  FIND_PACKAGE(Kokkos REQUIRED)
+  add_subdirectory(core)
+  add_subdirectory(containers)
+  add_subdirectory(algorithms)
+ELSE()
+
+
 # The build environment setup goes in the following steps
 # 1) Check all the enable options. This includes checking Kokkos_DEVICES
 # 2) Check the compiler ID (type and version)
@@ -279,3 +290,6 @@ IF (HAS_PARENT)
     SET_PROPERTY(GLOBAL PROPERTY Kokkos_ENABLE_${DEV} ON)
   ENDFOREACH()
 ENDIF()
+
+# End for not INSTALL_TESTING
+ENDIF()
```

And that seems to work properly. You only need to set compiler and Kokkos_INSTALL_TESTING  and everything else comes from the installed Kokkos instead of first trying to do a configuration of Kokkos and overwriting those settings via a find package.

**Comment by masterleinad** [comment]
Creating a separate directory with
```
cmake_minimum_required(VERSION 3.10)
project(KokkosTestInstall CXX)

find_package(Kokkos 3.1 REQUIRED QUIET)

SET(PACKAGE_NAME KOKKOS)
SET(KOKKOS_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/..)
SET(KOKKOS_ENABLE_TESTS ON)
SET(KOKKOS_INSTALL_TESTING ON)

include(${PROJECT_SOURCE_DIR}/../cmake/fake_tribits.cmake)
include(${PROJECT_SOURCE_DIR}/../cmake/kokkos_tribits.cmake)

ADD_SUBDIRECTORY(${PROJECT_SOURCE_DIR}/../core/ core)
ADD_SUBDIRECTORY(${PROJECT_SOURCE_DIR}/../algorithms algorithms)
ADD_SUBDIRECTORY(${PROJECT_SOURCE_DIR}/../containers containers)
```
would an alternative to be sure to not mess with the configuration of the found package.

**Comment by jjwilke** [comment]
I pushed a refactor that combines all the recent feedback (I think). 
* It receives the configuration from the imported package
* It skips any local configuration
* It uses an internal cache variable for the configuration
I kept the name `Kokkos_INSTALL_TESTING`. This seems a more accurate description than `DISABLE_COMPILING`?

**Comment by crtrott** [comment]
Yeah this looks good. Let me push again the Jenkins config file change I added which got nuked by your force push :-)

**Comment by masterleinad** [comment]
> I kept the name `Kokkos_INSTALL_TESTING`. This seems a more accurate description than `DISABLE_COMPILING`?

Now, `Kokkos_INSTALL_TESTING` implies `Kokkos_ENABLE_TESTS` which it didn't before IIRC. The name is fine with me then.
I'm a bit torn between handling all the configuration stuff in `CMakeLists.txt` and having a separate project (which is probably closer to what a user would have). Anyway, if both of you are in favor of doing it inline, that's OK.

**Comment by crtrott** [comment]
I think this is fine. I am also happy with the Kokkos_INSTALL_TESTING directly implying Kokkos_ENABLE_TESTS. That INSTALL_TESTING option anyway implies doing something completely different.

**Comment by crtrott** [comment]
And the install test worked in Jenkins. Great!

**Pull request merged**